test/tests/functional/pbs_power_provisioning_cray.py

```diff
@@ -449,15 +500,536 @@ def test_last_used_time(self):
         sts_time2 = int(time.mktime(time.strptime(fmttime, pattern)))
         self.assertGreater(sts_time2, sts_time1)
+
+    @timeout(1000)
+    def test_power_off_nodes(self):
+        """
+        Test power hook will power off the nodes if power_on_off_enable when
+        poweroff_eligible is set to true on nodes.
+        """
+        for n in self.server.status(NODE):
+            if 'resources_available.PBScraynid' in n:
+                self.server.manager(MGR_CMD_SET, NODE,
+                                    {"poweroff_eligible": True}, n['id'])
+        a = {"power_on_off_enable": True,
+             "max_concurrent_nodes": "30", 'node_idle_limit': '30'}
+        self.modify_hook_config(attrs=a, id='PBS_power')
+        a = {'freq': 30}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        a = {'enabled': 'True'}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        t = int(time.time())
+        self.logger.info("Waiting for 4 mins to power off all the nodes")
+        time.sleep(240)
+        self.server.log_match(
+            "/opt/cray/capmc/default/bin/capmc node_off", starttime=t)
+        for n in self.server.status(NODE):
+            if 'resources_available.PBScraynid' in n:
+                self.server.expect(NODE, {'state': 'sleep'}, id=n['id'])
+        self.cleanup_power_on()
+
+    @timeout(1000)
+    def test_power_on_off_max_concurennt_nodes(self):
+        """
+        Test power hook will power off the only the number of
+        max_concurrent nodes specified in conf file per hook run
+        even when poweroff_eligible is set to true on all the nodes.
+        """
+        for n in self.server.status(NODE):
+            if 'resources_available.PBScraynid' in n:
+                self.server.manager(MGR_CMD_SET, NODE,
+                                    {"poweroff_eligible": True}, n['id'])
+        a = {"power_on_off_enable": True, 'node_idle_limit': '10',
+             'max_concurrent_nodes': '2'}
+        self.modify_hook_config(attrs=a, id='PBS_power')
+        a = {'freq': 30}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        a = {'enabled': 'True'}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        self.logger.info("Waiting for 40 secs to power off 2 nodes")
+        time.sleep(40)
+        a = {'enabled': 'False'}
+        self.server.manager(MGR_CMD_SET, PBS_HOOK, a, id='PBS_power',
+                            sudo=True)
+        i = 0
+        for n in self.server.status(NODE):
+            if 'resources_available.PBScraynid' in n:
+                if n['state'] == 'sleep':
+                    i = i + 1
+        print i
+        if i != 2:
+            self.assertFalse("2 nodes are not in sleep state")
```

Comment:
I suggest you use self.server.expect(NODE, {'state=sleep': 2}) or use list comprehension to create your list and do a self.assertEquals(len(node_list), 2).  The list comprehension would be node_list = [n['id'] for n in self.server.status(NODE) if n['state'] == 'sleep']